enhancements/20231113-witness-node-support.md

```diff
@@ -0,0 +1,93 @@
+# Support witness node
+
+## Summary
+
+The witness node is a lightweight node that only runs the control plane components. The witness node is not schedulable and does not run any workload. The witness node forms a quorum with the other two nodes.
+
+### Related Issues
+
+https://github.com/harvester/harvester/issues/3266
+
+## Motivation
+
+Some users do not have three powerful nodes for the edge uses case. Instead, they would have two powerful nodes and one lightweight node. The lightweight node is only used for the `etcd` role. Kubernetes need at least three nodes to form a quorum. The witness node support will help the edge uses case. That will benefit both cost and high availability.
+
+### Goals
+
+Create a three-node cluster with one witness node.
+
+## Proposal
+
+### User Stories
+
+For the edge use case, the witness node support benefits both cost and high availability. The classical scenario would be two nodes and one witness node.
+
+### User Experience In Detail
+
+1. Set the node label `node-role.harvesterhci.io/etcd=true` to the node.
+2. The promoter will add the node to the cluster as `etcd` role.
+
+### API changes
+
+None
+
+## Design
+
+### Limitations
+
+- Each cluster only supports one witness node.
+
+### Implementation Overview
+
+#### Promote controller
+
+Harvester depends on the job `promote` to promote a node to related role after it joins (the default role is worker). We can only promote the node to `etcd` role with the label `node-role.harvesterhci.io/etcd=true`. Also, we need to add taint for etcd role to prevent workload from scheduling to the node.
```

Comment:
Before the `promote` is finished, could some workload be scheduled to this `witness-node`? As it joins the cluster as `worker` from beginning, k8s may schedule any none-control related workload to this node at any time.

Namely, the `taint` needs to be done at very early stage.

Reply:
Yap, that's a good point.
In fact, as a worker node, we indeed deploy a few workloads on it after installation (like rancher/harvester pod)
We will set taint before promoting the operation, which will delete the unneeded pod on this node.

I do not think that will have too much impact on the above workloads, but if needed, we could try to add taint at a very early stage.